Create design-qa-considerations.md (#26774)
fleetdm/fleet: docs/Contributing/design-qa-considerations.md

# Design / QA considerations

This is meant to be a helpful checklist of 'events' or 'transactions' to remember to think about while designing or testing new features or bugs to make sure edge cases are caught sooner rather than later. Please feel free to add more if any are missing

## User

> - Create User

Reviewer: ```suggestion
- Create user
```

Looking good! I think let's update this to use sentence case everywhere: https://fleetdm.com/handbook/company/why-this-way#why-does-fleet-use-sentence-case

## MDM

- Turn MDM On
- Turn MDM Off
- Enable Disk Encryption
- Disable Disk Encryption
- Add ABM Token
- Add Multiple ABM Tokens
- Remove ABM Token
- Add VPP Token
- Add Multiple VPP Tokens
- Remove VPP Token
- Add minimum version OS Updates
- Remove minimum version OS Updates
- Add Profile
- Remove Profile
- Resend Profile
- Add Bootstrap Package
- Remove Bootstrap Package
- Single Host Turn MDM Off
- Setup Experience Software / scripts

## Software

- Add Software to team
  - FMA
  - VPP
  - Custom Package
- Remove Software from team
- Edit Software (Scripts / binary)
- Add Script
- Run Script
- Edit Script
- Remove Script
- Vulnerability Scans
- Automatic Software install
- Label-scoped software install

## Policy

- Add Policy
- Remove Policy
- Add install automation
- Add Calendar automation

## Query

- Add Query
- Remove Query
- Edit Query
- Live Query
- Saved Query Results

## Labels

- Add Dynamic Label
- Remove Dynamic Label
- Add Manual Label
- Remove Manual Label
- Add host to an existing label
- Remove a host from a label
- Label selecion (policy / profile / software)
  - Include all
  - Include Any
  - Exclude All
  - Exclude Any

## Host

> - Enroll to fleet from package

Reviewer: ```suggestion
- Enroll to Fleet using Fleet's agent (fleetd)
```

- Enrolled via osquery (no orbit / fleetd)

> - Deleted from fleet

Reviewer: ```suggestion
- Deleted from Fleet
```

Capital "F" for "Fleet." Looks like we have lower case "f" in a couple places

# Design / QA considerations

> This is meant to be a helpful checklist of 'events' or 'transactions' to remember to think about while designing or testing new features or bugs to make sure edge cases are caught sooner rather than later. Please feel free to add more if any are missing

Reviewer 2: ```suggestion
This is meant to be a helpful checklist of 'events' or 'transactions' to help catch edge cases sooner rather than later while designing or testing new features or bugs.  Please feel free to add more if any are missing.
```

- EULA added for DEP enrollment

## Software

- Add software to team
  - FMA
  - VPP
  - custom package
- Remove software from team
- Edit software (scripts / binary)
- Add script
- Run script
- Edit script
- Remove script
- Vulnerability scans
- Automatic software install
- Label-scoped software install

## Policy

- Add policy
- Remove policy
- Add install automation
- Add Calendar automation

## Query

- Add query
- Remove query
- Edit query
- Live query
- Saved query results

## Labels

- Add dynamic label
- Remove dynamic label
- Add Manual label
- Remove Manual label
- Add host to an existing label
- Remove a host from a label
- Label selecion (policy / profile / software)
  - Include all
  - Include any
  - Exclude all
  - Exclude any

## Host

- Enroll to Fleet using Fleet's agent (fleetd)
- Enrolled via osquery (no orbit / fleetd)
- Deleted from Fleet
- DEP enrollment
- BYOD enrollment
- ABM ghost host before enrolled
- Wiped host
- Locked host
- Host that succeeds all policies
- Host with a failing policy
- Online host
- Offline host

## Integrations

- Jira
- Zendesk
- Webhooks

## Config

- Host callback times other than 1hr
- DB primary / replica

> - Async ingestion of policies

Reviewer 2: ```suggestion
- Async ingestion of policies

<meta name="pageOrderInSection" value="3300">
<meta name="description" value="A helpful checklist of 'events' or 'transactions' to think about while designing or testing new features or bugs.">
```

# Design / QA considerations

This is meant to be a helpful checklist of 'events' or 'transactions' to remember to think about while designing or testing new features or bugs to make sure edge cases are caught sooner rather than later. Please feel free to add more if any are missing

## User

- Create User
- Remove User

> - Update User permissions

Reviewer 3: Create API-only user (special-case)

## Team

- Create Team
- Remove Team
- Transfer Host into this team

> - Transfer Host out of this team

Reviewer 3: Not an "event", but I think it's important to call out "No team" vs "All teams" here (unless it's already somewhere else, having scrolled all the way yet).

## MDM

- Turn MDM On
- Turn MDM Off
- Enable Disk Encryption
- Disable Disk Encryption
- Add ABM Token
- Remove ABM Token
- Add VPP Token

> - Remove VPP Token

Reviewer 3: Worth calling out "Add multiple ABM tokens"/"Add multiple VPP tokens" just to ensure we think through the uncommon case where there are more than one.

- Add minimum version OS Updates
- Remove minimum version OS Updates
- Add Profile
- Remove Profile
- Add Bootstrap Package
- Remove Bootstrap Package

> - Setup Experience Software / scripts

Reviewer 3: Resend profile, turn MDM on manually/BYOD, turn MDM off for a single host.

## Software

> - Add Software to team

Reviewer 3: Just to make it explicit so folks think about all possibilities, "Add FMA/VPP/Custom Software to team", "Remove /Edit...".

- Remove Software from team
- Edit Software (Scripts / binary)
- Add Script

> - Run Script

Reviewer 3: Sync/Async

Author: Sync has been updated to just call async and watch the pending job until completion. So I think we can just talk about async really atm. If we add more sync I would say add this split for sure.

- Edit Script
- Remove Script

> - Vulnerability Scans

Reviewer 3: Automatic software install, label-scoped software install (include/exclude all/any).

## Policy

- Add Policy
- Remove Policy
- Add install automation
- Add Calendar automation

## Query

- Add Query
- Remove Query

> - Edit Query

Reviewer 3: Live query, and something about saved query results (not quite sure myself what/where that is).

## Labels

> - Add Automatic Label

Reviewer 3: ```suggestion
- Add Dynamic Label
```

- Remove Automatic Label
- Add Manual Label

> - Remove Manual Label

Reviewer 3: Add/Remove hosts to/from (manual) label.

## Host

> - Enroll to fleet from package

Reviewer 3: Enroll with/without orbit/fleetd

- Deleted from fleet
- DEP Enrollment

> - BYOD Enrollment

Reviewer 3: Is that how we call all manual MDM enrollments now?

Author: it's becoming pretty overloaded to at this point I would say

- ABM ghost host before enrolled
- Wiped host
- Locked host
- Host that succeeds all policies
- Host with a failing policy
- Online Host

> - Offline Host

Reviewer 3: I'd add a couple sections:

### Integrations

- Jira
- Zendesk
- Webhooks

### Config 
Not "events" at all, but some special cases worth calling out?

- All special/unusual things that can be done with the config
- Maybe calling out that host callback can be configured, not necessarily 1h
- DB primary/replica scenarios and potential lag
- The async ingestion of policies and other stuff (can't remember them all but I think there are 3-4 things that can be ingested async for - significant IIRC - extra performance)
